music/src/core/soundfont.ts

```diff
@@ -0,0 +1,377 @@
+ * always be played in its entirety.
+ * @param velocities (Optional) The set of velocities sampled.
+ */
+export interface InstrumentSpec {
+  name: string;
+  minPitch: number;
+  maxPitch: number;
+  durationSeconds: number;
+  releaseSeconds: number;
+  percussive: boolean;
+  velocities?: number[];
+}
+
+/**
+ * Sampled instrument. Must be initialized and samples must be pre-loaded using
+ * the `loadSamples` method before any notes can be played.
+ */
+export class Instrument {
+  private FADE_SECONDS = 0.1;
+
+  private readonly baseURL: string;
+  // tslint:disable-next-line:no-any
+  private readonly buffers: any;
+
+  private initialized: boolean;
+
+  name: string;
+  minPitch: number;
+  maxPitch: number;
+  durationSeconds: number;
+  releaseSeconds: number;
+  percussive: boolean;
+  velocities?: number[];
+
+  /**
+   * `Instrument` constructor.
+   *
+   * @param baseURL Path to the instrument directory.
+   */
+  constructor(baseURL: string) {
+    this.baseURL = baseURL;
+    this.buffers = new Tone.Buffers([]);
+    this.initialized = false;
+  }
+
+  /**
+   * Loads instrument configuration from an `instrument.json` file in the base
+   * URL directory. Does not load any of the samples.
+   */
+  async initialize() {
+    await fetch(`${this.baseURL}/instrument.json`)
+        .then((response) => response.json())
+        .then((spec: InstrumentSpec) => {
+          this.name = spec.name;
+          this.minPitch = spec.minPitch;
+          this.maxPitch = spec.maxPitch;
+          this.durationSeconds = spec.durationSeconds;
+          this.releaseSeconds = spec.releaseSeconds;
+          this.percussive = spec.percussive;
+          this.velocities = spec.velocities;
+          this.initialized = true;
+        });
+  }
+
+  /**
+   * Map pitch and velocity to sample name.
+   */
+  private sampleInfoToName(sampleInfo: SampleInfo) {
+    if (this.velocities) {
+      return `p${sampleInfo.pitch}_v${sampleInfo.velocity}`;
+    } else {
+      return `p${sampleInfo.pitch}`;
+    }
+  }
+
+  /**
+   * Map pitch and velocity to sample URL.
+   */
+  private sampleInfoToURL(sampleInfo: SampleInfo) {
+    return `${this.baseURL}/${this.sampleInfoToName(sampleInfo)}.mp3`;
+  }
+
+  /**
+   * Find nearest sampled velocity to a target velocity.
+   */
+  private nearestVelocity(velocity: number): number {
+    if (!this.velocities) {
+      return velocity;
+    }
+
+    let bestVelocity = undefined;
+    let bestDistance = constants.MIDI_VELOCITIES;
+    this.velocities.forEach((v) => {
+      const d = Math.abs(v - velocity);
+      if (d < bestDistance) {
+        bestVelocity = v;
+        bestDistance = d;
+      }
+    });
+    return bestVelocity;
+  }
+
+  /**
+   * Load samples necessary to play a set of pitch/velocity pairs. This must be
+   * called before any notes can be played.
+   *
+   * @param samples Array of pitch/velocity pairs.
+   */
+  async loadSamples(samples: SampleInfo[]) {
```

Comment:
does this avoid reloading samples?

Reply:
Yes, it filters out samples for which we already have a buffer.  Will add a comment.